odbc/src/api/functions.rs

```diff
@@ -1312,73 +1307,235 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
+                // Additionally, rows_fetch_ptr is set to 0.
+                return SqlReturn::NO_DATA;
+            }
+
+            has_fetched_at_least_one_row = true;
+
+            if has_row_status_array {
+                *row_status_buffer = if row_warnings_opt.is_empty() {
+                    RowStatus::SQL_ROW_SUCCESS as USmallInt
+                } else {
+                    RowStatus::SQL_ROW_SUCCESS_WITH_INFO as USmallInt
+                };
+            }
+
+            // keep track of all warnings that occur throughout the entire function.
+            global_fetch_warnings_opt.append(&mut row_warnings_opt);
+
+            fetched_rows += 1;
+
+            *stmt.var_data_cache.write().unwrap() = Some(HashMap::new());
+
+            // If there are bound columns, then copy data from the result set into the bound buffers.
+            if let Some(bound_cols) = stmt.bound_cols.read().unwrap().as_ref() {
+                let (
+                    encountered_error_during_col_binding,
+                    encountered_success_with_info_during_col_binding,
+                ) = sql_fetch_bound_buffers(
+                    statement_handle,
+                    index,
+                    row_status_buffer,
+                    bound_cols,
+                    function_name,
+                );
+
+                // It's possible that one binding causes SUCCESS_WITH_INFO and another causes ERROR, so we need to check for an error first
+                // and ignore SUCCESS_WITH_INFO if it also occurs.
+                if encountered_error_during_col_binding {
+                    row_error_count += 1;
+                }
+                // keep track if SUCCESS_WITH_INFO is ever encountered throughout the entire function.
+                else if encountered_success_with_info_during_col_binding {
+                    encountered_success_with_info = true;
                 }
             }
-        }
-        if !warnings_opt.is_empty() || success_with_info_encountered {
-            // No warnings and there is a next row
-            SqlReturn::SUCCESS_WITH_INFO
         } else {
-            SqlReturn::SUCCESS
+            // An error happened when moving the cursor and fetching the next row
+            let mongo_handle = MongoHandleRef::from(statement_handle);
+            add_diag_with_function!(
+                mongo_handle,
+                move_to_next_result.as_ref().unwrap_err().clone(),
+                function_name.to_string()
+            );
+
+            let error = move_to_next_result.unwrap_err();
+
+            // Checks if there is an error that applies to the entire function instead of just one row.
+            // If there is, early exit with SqlReturn::Error.
+            if matches!(
+                error,
+                ODBCError::InvalidCursorState | ODBCError::Core(Error::QueryCancelled)
+            ) {
+                return SqlReturn::ERROR;
+            }
```

Comment:
This is what I made so far to check for errors that apply to the entire function and require an early exit. What other errors should I be checking for?